File: src/lib/utils/native-web-search.ts
import type { Model, NativeWebSearchSupport } from '$lib/stores';
import { normalizeWebSearchMode, type WebSearchMode } from '$lib/utils/web-search-mode';
import nativeWebSearchRules from '$lib/data/native-web-search-rules.json';

type Translator = (key: string, options?: Record<string, unknown>) => string;

type WebSearchConfigLike = {
	features?: {
		enable_web_search?: boolean;
		enable_halo_web_search?: boolean;
		enable_native_web_search?: boolean;
		default_web_search_mode?: string;
	};
};

type ModelLike = Partial<Model> & {
	id?: string;
	name?: string;
	owned_by?: string;
	native_web_search_supported?: boolean;
	native_web_search_support?: NativeWebSearchSupport;
};

export type WebSearchModeOption = {
	value: WebSearchMode;
	label: string;
	shortLabel?: string;
	description?: string;
	descriptionTone?: 'default' | 'info' | 'warning';
	disabled?: boolean;
	badge?: string;
};

export type NativeWebSearchSummary = {
	total: number;
	supportedCount: number;
	unknownCount: number;
	unsupportedCount: number;
	anySupported: boolean;
	anyUnknown: boolean;
	anyUnsupported: boolean;
	allSupported: boolean;
	allUnsupported: boolean;
	hasSelection: boolean;
	sampleSupported?: NativeWebSearchSupport;
	sampleUnknown?: NativeWebSearchSupport;
	sampleUnsupported?: NativeWebSearchSupport;
};

const SUPPORTED_STATUSES = new Set<NativeWebSearchSupport['status']>([
	'supported',
	'unknown',
	'unsupported'
]);

type NativeWebSearchRule = {
	type?: string;
	value?: string;
	reason?: string;
};

type NativeWebSearchProviderRules = {
	default_status?: NativeWebSearchSupport['status'];
	allow?: NativeWebSearchRule[];
	deny?: NativeWebSearchRule[];
};

const RULE_MATCH_REGEX = 'regex';
const RULE_MATCH_CONTAINS = 'contains';
const RULE_MATCH_EQUALS = 'equals';
const RULE_MATCH_PREFIX = 'prefix';

function normalizeRuleProvider(provider?: string | null): string {
	const normalized = (provider ?? '').toString().trim().toLowerCase();
	if (normalized === 'google' || normalized === 'gemini') {
		return 'gemini';
	}
	return normalized || 'unknown';
}

function normalizeModelLookupValue(value?: string | null): string {
	const normalized = (value ?? '').toString().trim().toLowerCase();
	if (!normalized) return '';
	return normalized.startsWith('models/') ? normalized.slice('models/'.length) : normalized;
}

function getProviderRules(provider?: string | null): NativeWebSearchProviderRules {
	const providers = (nativeWebSearchRules as { providers?: Record<string, NativeWebSearchProviderRules> })
		.providers;
	if (!providers || typeof providers !== 'object') {
		return {};
	}
	return providers[normalizeRuleProvider(provider)] ?? {};
}

function matchesRule(rule: NativeWebSearchRule, candidate: string): boolean {
	const matchType = (rule?.type ?? '').toString().trim().toLowerCase();
	const ruleValue = (rule?.value ?? '').toString().trim();

	if (!candidate || !matchType || !ruleValue) {
		return false;
	}

	switch (matchType) {
		case RULE_MATCH_REGEX:
			return new RegExp(ruleValue, 'i').test(candidate);
		case RULE_MATCH_CONTAINS:
			return candidate.includes(ruleValue.toLowerCase());
		case RULE_MATCH_EQUALS:
			return candidate === ruleValue.toLowerCase();
		case RULE_MATCH_PREFIX:
			return candidate.startsWith(ruleValue.toLowerCase());
		default:
			return false;
	}
}

function resolveFallbackModelRule(
	provider?: string | null,
	model?: Pick<ModelLike, 'id' | 'name'> | null
): NativeWebSearchSupport {
	const normalizedProvider = normalizeRuleProvider(provider);
	const providerRules = getProviderRules(normalizedProvider);
	if (!providerRules.default_status && normalizedProvider !== 'openai' && normalizedProvider !== 'gemini') {
		return {
			status: 'unsupported',
			reason: normalizedProvider ? 'provider_not_supported' : 'unknown_model',
			source: 'inferred',
			provider: normalizedProvider,
			supported: false,
			can_attempt: false
		};
	}
	const candidates = [
		normalizeModelLookupValue(model?.id ?? ''),
		normalizeModelLookupValue(model?.name ?? '')
	].filter(Boolean);

	for (const [group, status] of [
		['deny', 'unsupported'],
		['allow', 'supported']
	] as const) {
		const rules = Array.isArray(providerRules[group]) ? providerRules[group] : [];
		for (const rule of rules) {
			for (const candidate of candidates) {
				if (matchesRule(rule, candidate)) {
					return {
						status,
						reason: rule.reason ?? 'model_rule_unknown',
						source: 'model_rules_fallback',
						provider: normalizedProvider,
						supported: status === 'supported',
						can_attempt: status !== 'unsupported',
						model_rule: {
							status,
							reason: rule.reason ?? 'model_rule_unknown',
							source: 'model_rules_fallback',
							match_type: rule.type,
							match_value: rule.value,
							matched_on: candidate
						}
					};
				}
			}
		}
	}

	return {
		status: providerRules.default_status ?? 'unknown',
		reason: 'model_rule_unknown',
		source: 'model_rules_fallback',
		provider: normalizedProvider,
		supported: false,
		can_attempt: normalizedProvider === 'openai' || normalizedProvider === 'gemini',
		model_rule: {
			status: providerRules.default_status ?? 'unknown',
			reason: 'model_rule_unknown',
			source: 'model_rules_fallback'
		}
	};
}

export function getNativeWebSearchSupport(model?: ModelLike | null): NativeWebSearchSupport {
	const support = model?.native_web_search_support;
	if (support && typeof support === 'object' && SUPPORTED_STATUSES.has(support.status)) {
		return support;
	}

	const fallbackRuleSupport = resolveFallbackModelRule(model?.owned_by, model);

	if (model?.native_web_search_supported === true) {
		return {
			status: 'supported',
			reason: 'legacy_supported',
			source: 'legacy'
		};
	}

	if (model?.native_web_search_supported === false) {
		return {
			...fallbackRuleSupport,
			status: 'unsupported',
			reason: fallbackRuleSupport.reason === 'model_rule_unknown' ? 'legacy_unsupported' : fallbackRuleSupport.reason,
			source: fallbackRuleSupport.source === 'model_rules_fallback' ? 'legacy+model_rules_fallback' : 'legacy'
		};
	}

	if (fallbackRuleSupport.source === 'model_rules_fallback') {
		return fallbackRuleSupport;
	}

	return {
		status: 'unsupported',
		reason: (model?.owned_by ?? '').toString().trim() ? 'provider_not_supported' : 'unknown_model',
		source: 'inferred'
	};
}

export function summarizeNativeWebSearchSupport(models: Array<ModelLike | null | undefined>): NativeWebSearchSummary {
	const validModels = models.filter(Boolean) as ModelLike[];
	const summary: NativeWebSearchSummary = {
		total: validModels.length,
		supportedCount: 0,
		unknownCount: 0,
		unsupportedCount: 0,
		anySupported: false,
		anyUnknown: false,
		anyUnsupported: false,
		allSupported: false,
		allUnsupported: false,
		hasSelection: validModels.length > 0
	};

	for (const model of validModels) {
		const support = getNativeWebSearchSupport(model);
		if (support.status === 'supported') {
			summary.supportedCount += 1;
			summary.sampleSupported ??= support;
			continue;
		}
		if (support.status === 'unknown') {
			summary.unknownCount += 1;
			summary.sampleUnknown ??= support;
			continue;
		}
		summary.unsupportedCount += 1;
		summary.sampleUnsupported ??= support;
	}

	summary.anySupported = summary.supportedCount > 0;
	summary.anyUnknown = summary.unknownCount > 0;
	summary.anyUnsupported = summary.unsupportedCount > 0;
	summary.allSupported = summary.hasSelection && summary.supportedCount === summary.total;
	summary.allUnsupported = summary.hasSelection && summary.unsupportedCount === summary.total;

	return summary;
}

export function describeNativeWebSearchSupport(
	t: Translator,
	support?: NativeWebSearchSupport | null
): string {
	switch (support?.reason) {
		case 'official_connection':
			return t('Official provider endpoint detected. This model can use built-in web search.');
		case 'connection_enabled':
			return t('Built-in web search is enabled for this connection.');
		case 'connection_disabled':
			return t('Built-in web search is disabled for this connection.');
		case 'compat_connection_unverified':
			return t(
				'This compatible endpoint is not verified yet. If the upstream supports built-in search tools, you can still try model-native web search in chat.'
			);
		case 'provider_not_supported':
			return t('This provider does not expose model-native web search in HaloWebUI yet.');
		case 'connection_not_found':
			return t('HaloWebUI could not resolve the connection behind this model.');
		default:
			break;
	}

	if (support?.status === 'supported') {
		return t('Model-native web search is available for this model.');
	}
	if (support?.status === 'unknown') {
		return t('Native web search availability for this model is currently unknown.');
	}
	return t('Model-native web search is unavailable for this model.');
}

export function getNativeWebSearchAvailabilityNote(
	t: Translator,
	summary: NativeWebSearchSummary,
	scope: 'selection' | 'catalog' = 'selection'
): string {
	if (!summary.hasSelection) {
		return '';
	}

	const prefix =
		scope === 'selection' ? t('Current selection') : t('Currently loaded models');

	if (summary.allSupported) {
		return t('{{scope}}: all {{count}} models support native web search.', {
			scope: prefix,
			count: summary.total
		});
	}

	if (summary.allUnsupported) {
		return t('{{scope}}: native web search is unavailable for all models.', {
			scope: prefix
		});
	}

	return t('{{scope}}: {{supported}} native, {{unknown}} unverified, {{unsupported}} unavailable.', {
		scope: prefix,
		supported: summary.supportedCount,
		unknown: summary.unknownCount,
		unsupported: summary.unsupportedCount
	});
}

function buildNativeModeDescription(t: Translator, summary: NativeWebSearchSummary): string {
	if (summary.allUnsupported) {
		return t('Model-native web search is unavailable for this model.');
	}

	return t('使用模型自带的联网搜索。');
}

function buildAutoModeDescription(
	t: Translator,
	haloEnabled: boolean,
	summary: NativeWebSearchSummary
): string {
	if (summary.hasSelection && !haloEnabled && !summary.anySupported) {
		return t('No web search route is available for the current selection.');
	}

	if (summary.allUnsupported && !haloEnabled) {
		return t('Model-native web search is unavailable for this model.');
	}
	if (summary.supportedCount === 0 && summary.unknownCount > 0 && !haloEnabled) {
		return t(
			'No automatic web search route is available for the current selection. You can still try model-native web search manually.'
		);
	}

	return t('自动判断是否需要联网，适合日常使用。');
}

export function getSmartWebSearchRouteLabel(
	t: Translator,
	config: WebSearchConfigLike | null | undefined,
	models: Array<ModelLike | null | undefined>
): string {
	const haloEnabled = Boolean(
		config?.features?.enable_halo_web_search ?? config?.features?.enable_web_search
	);
	const nativeEnabled = Boolean(config?.features?.enable_native_web_search);
	const summary = summarizeNativeWebSearchSupport(models);

	if (nativeEnabled && summary.anySupported) {
		return t('Smart · Model Native');
	}

	if (haloEnabled) {
		return t('Smart · HaloWebUI');
	}

	return t('Smart');
}

export function buildWebSearchModeOptions(
	t: Translator,
	config: WebSearchConfigLike | null | undefined,
	models: Array<ModelLike | null | undefined>
): WebSearchModeOption[] {
	const haloEnabled = Boolean(
		config?.features?.enable_halo_web_search ?? config?.features?.enable_web_search
	);
	const nativeEnabled = Boolean(config?.features?.enable_native_web_search);
	const summary = summarizeNativeWebSearchSupport(models);
	const nativeImpossible = summary.hasSelection && summary.allUnsupported;
	const autoImpossible = summary.hasSelection && !haloEnabled && !summary.anySupported;
	const nativeDescriptionTone: WebSearchModeOption['descriptionTone'] = summary.allUnsupported
		? 'warning'
		: summary.supportedCount === 0 && summary.unknownCount > 0
			? 'info'
			: 'default';
	const autoDescriptionTone: WebSearchModeOption['descriptionTone'] =
		summary.allUnsupported || (summary.supportedCount === 0 && summary.unknownCount > 0)
			? 'info'
			: 'default';

	return [
		{
			value: 'off',
			label: t('关闭联网'),
			description: t('本次对话不联网。')
		},
		...(haloEnabled
			? [
					{
						value: 'halo' as WebSearchMode,
						label: 'HaloWebUI 搜索',
						description: t('使用 HaloWebUI 搜索网页。')
					}
				]
			: []),
		...(nativeEnabled
			? [
					{
						value: 'native' as WebSearchMode,
						label: t('模型原生联网'),
						description: buildNativeModeDescription(t, summary),
						descriptionTone: nativeDescriptionTone,
						disabled: nativeImpossible
					}
				]
			: []),
		...(haloEnabled || nativeEnabled
			? [
					{
						value: 'auto' as WebSearchMode,
						label: t('Smart Web Search'),
						shortLabel: getSmartWebSearchRouteLabel(t, config, models),
						description: buildAutoModeDescription(t, haloEnabled, summary),
						descriptionTone: autoDescriptionTone,
						disabled: autoImpossible,
						badge: haloEnabled && !autoImpossible ? t('Recommended') : undefined
					}
				]
			: [])
	];
}

export function resolveConfiguredDefaultWebSearchMode(
	t: Translator,
	config: WebSearchConfigLike | null | undefined,
	models: Array<ModelLike | null | undefined>,
	canUseWebSearch: boolean
): WebSearchMode {
	if (!canUseWebSearch) {
		return 'off';
	}

	const configuredMode = normalizeWebSearchMode(config?.features?.default_web_search_mode, 'off');
	if (configuredMode === 'off') {
		return 'off';
	}

	const availableModes = buildWebSearchModeOptions(t, config, models);
	return availableModes.some(
		(option) => option.value === configuredMode && option.disabled !== true
	)
		? configuredMode
		: 'off';
}

File: backend/open_webui/utils/native_web_search.py
import json
import re
from functools import lru_cache
from pathlib import Path
from typing import Optional
from urllib.parse import urlparse


NATIVE_WEB_SEARCH_STATUS_SUPPORTED = "supported"
NATIVE_WEB_SEARCH_STATUS_UNKNOWN = "unknown"
NATIVE_WEB_SEARCH_STATUS_UNSUPPORTED = "unsupported"
NATIVE_WEB_SEARCH_EFFECTIVE_SCOPE = "connection+model"

RULE_MATCH_REGEX = "regex"
RULE_MATCH_CONTAINS = "contains"
RULE_MATCH_EQUALS = "equals"
RULE_MATCH_PREFIX = "prefix"

PROJECT_ROOT = Path(__file__).resolve().parents[3]
NATIVE_WEB_SEARCH_RULES_PATH = (
    PROJECT_ROOT / "src" / "lib" / "data" / "native-web-search-rules.json"
)


def _get_hostname(value: str) -> str:
    try:
        return (urlparse(value).hostname or "").strip().lower()
    except Exception:
        return ""


def is_official_openai_connection(url: str) -> bool:
    host = _get_hostname(url)
    return host == "api.openai.com" or host.endswith(".openai.com")


def is_official_gemini_connection(url: str) -> bool:
    host = _get_hostname(url)
    return host == "generativelanguage.googleapis.com"


def _normalize_rule_provider(provider: str) -> str:
    normalized = str(provider or "").strip().lower()
    if normalized in {"google", "gemini"}:
        return "gemini"
    return normalized or "unknown"


def _normalize_model_lookup_value(value: str) -> str:
    normalized = str(value or "").strip().lower()
    if not normalized:
        return ""
    if normalized.startswith("models/"):
        normalized = normalized[len("models/") :]
    return normalized


def strip_model_prefix(model_id: str, prefix_id: Optional[str]) -> str:
    normalized_id = str(model_id or "").strip()
    prefix = str(prefix_id or "").strip()
    if prefix and normalized_id.startswith(f"{prefix}."):
        return normalized_id[len(prefix) + 1 :]
    return normalized_id


@lru_cache(maxsize=1)
def load_native_web_search_rules() -> dict:
    try:
        with NATIVE_WEB_SEARCH_RULES_PATH.open("r", encoding="utf-8") as file:
            data = json.load(file)
        if isinstance(data, dict):
            return data
    except Exception:
        pass
    return {"providers": {}}


def _get_provider_rules(provider: str) -> dict:
    rules = load_native_web_search_rules()
    providers = rules.get("providers", {})
    if not isinstance(providers, dict):
        return {}
    value = providers.get(_normalize_rule_provider(provider), {})
    return value if isinstance(value, dict) else {}


def _match_rule(rule: dict, candidate: str) -> bool:
    if not candidate or not isinstance(rule, dict):
        return False

    match_type = str(rule.get("type") or "").strip().lower()
    rule_value = str(rule.get("value") or "").strip()
    if not match_type or not rule_value:
        return False

    if match_type == RULE_MATCH_REGEX:
        return re.search(rule_value, candidate, flags=re.IGNORECASE) is not None
    if match_type == RULE_MATCH_CONTAINS:
        return rule_value.lower() in candidate
    if match_type == RULE_MATCH_EQUALS:
        return candidate == rule_value.lower()
    if match_type == RULE_MATCH_PREFIX:
        return candidate.startswith(rule_value.lower())
    return False


def resolve_model_native_web_search_rule(
    provider: str,
    *,
    model_id: str = "",
    model_name: str = "",
) -> dict:
    normalized_provider = _normalize_rule_provider(provider)
    provider_rules = _get_provider_rules(normalized_provider)

    values = []
    for raw in (model_id, model_name):
        normalized = _normalize_model_lookup_value(raw)
        if normalized and normalized not in values:
            values.append(normalized)

    if not provider_rules:
        return {
            "provider": normalized_provider,
            "status": NATIVE_WEB_SEARCH_STATUS_UNKNOWN,
            "reason": "model_rule_unknown",
            "source": "model_rules",
        }

    for group_name, status in (
        ("deny", NATIVE_WEB_SEARCH_STATUS_UNSUPPORTED),
        ("allow", NATIVE_WEB_SEARCH_STATUS_SUPPORTED),
    ):
        group = provider_rules.get(group_name, [])
        if not isinstance(group, list):
            continue
        for rule in group:
            if not isinstance(rule, dict):
                continue
            for candidate in values:
                if _match_rule(rule, candidate):
                    return {
                        "provider": normalized_provider,
                        "status": status,
                        "reason": str(rule.get("reason") or "model_rule_unknown"),
                        "source": "model_rules",
                        "match_type": rule.get("type"),
                        "match_value": rule.get("value"),
                        "matched_on": candidate,
                    }

    default_status = str(provider_rules.get("default_status") or "").strip().lower()
    if default_status not in {
        NATIVE_WEB_SEARCH_STATUS_SUPPORTED,
        NATIVE_WEB_SEARCH_STATUS_UNKNOWN,
        NATIVE_WEB_SEARCH_STATUS_UNSUPPORTED,
    }:
        default_status = NATIVE_WEB_SEARCH_STATUS_UNKNOWN

    return {
        "provider": normalized_provider,
        "status": default_status,
        "reason": "model_rule_unknown",
        "source": "model_rules",
    }


def resolve_effective_native_web_search_support(
    connection_support: dict,
    *,
    provider: str,
    model_id: str = "",
    model_name: str = "",
) -> dict:
    connection = dict(connection_support or {})
    normalized_provider = _normalize_rule_provider(
        connection.get("provider") or provider
    )
    model_rule = resolve_model_native_web_search_rule(
        normalized_provider, model_id=model_id, model_name=model_name
    )

    connection_status = connection.get("status") or NATIVE_WEB_SEARCH_STATUS_UNKNOWN
    model_status = model_rule.get("status") or NATIVE_WEB_SEARCH_STATUS_UNKNOWN

    if connection_status == NATIVE_WEB_SEARCH_STATUS_UNSUPPORTED:
        effective_status = NATIVE_WEB_SEARCH_STATUS_UNSUPPORTED
        reason = connection.get("reason") or "connection_not_found"
        source = connection.get("source") or "connection_support"
    elif model_status == NATIVE_WEB_SEARCH_STATUS_UNSUPPORTED:
        effective_status = NATIVE_WEB_SEARCH_STATUS_UNSUPPORTED
        reason = model_rule.get("reason") or "model_rule_unknown"
        source = model_rule.get("source") or "model_rules"
    elif (
        connection_status == NATIVE_WEB_SEARCH_STATUS_SUPPORTED
        and model_status == NATIVE_WEB_SEARCH_STATUS_SUPPORTED
    ):
        effective_status = NATIVE_WEB_SEARCH_STATUS_SUPPORTED
        reason = model_rule.get("reason") or connection.get("reason") or "official_connection"
        source = NATIVE_WEB_SEARCH_EFFECTIVE_SCOPE
    elif (
        normalized_provider == "openai"
        and connection_status == NATIVE_WEB_SEARCH_STATUS_UNKNOWN
        and model_status == NATIVE_WEB_SEARCH_STATUS_SUPPORTED
    ):
        effective_status = NATIVE_WEB_SEARCH_STATUS_SUPPORTED
        reason = model_rule.get("reason") or "model_rule_allow_openai_compatible"
        source = NATIVE_WEB_SEARCH_EFFECTIVE_SCOPE
    else:
        effective_status = NATIVE_WEB_SEARCH_STATUS_UNKNOWN
        reason = model_rule.get("reason") or connection.get("reason") or "model_rule_unknown"
        source = NATIVE_WEB_SEARCH_EFFECTIVE_SCOPE

    can_attempt = (
        effective_status in {
            NATIVE_WEB_SEARCH_STATUS_SUPPORTED,
            NATIVE_WEB_SEARCH_STATUS_UNKNOWN,
        }
        and connection_status != NATIVE_WEB_SEARCH_STATUS_UNSUPPORTED
        and model_status != NATIVE_WEB_SEARCH_STATUS_UNSUPPORTED
    )

    return {
        **{
            k: v
            for k, v in connection.items()
            if k not in {"supported", "can_attempt"}
        },
        "provider": normalized_provider,
        "status": effective_status,
        "reason": reason,
        "source": source,
        "supported": effective_status == NATIVE_WEB_SEARCH_STATUS_SUPPORTED,
        "can_attempt": can_attempt,
        "connection_support": {
            k: v
            for k, v in connection.items()
            if k not in {"url", "api_config", "connection_support", "model_rule"}
        },
        "model_rule": model_rule,
        "effective_scope": NATIVE_WEB_SEARCH_EFFECTIVE_SCOPE,
    }


def build_native_web_search_support(
    provider: str,
    *,
    url: str = "",
    api_config: Optional[dict] = None,
    connection_name: Optional[str] = None,
) -> dict:
    normalized_provider = str(provider or "").strip().lower()
    config = api_config if isinstance(api_config, dict) else {}
    explicit = config.get("native_web_search_enabled")
    configured = explicit if isinstance(explicit, bool) else None

    if normalized_provider == "openai":
        official = is_official_openai_connection(url)
        if configured is True:
            status = NATIVE_WEB_SEARCH_STATUS_SUPPORTED
            reason = "connection_enabled"
            source = "connection_config"
        elif configured is False:
            status = NATIVE_WEB_SEARCH_STATUS_UNSUPPORTED
            reason = "connection_disabled"
            source = "connection_config"
        elif official:
            status = NATIVE_WEB_SEARCH_STATUS_SUPPORTED
            reason = "official_connection"
            source = "official_connection"
        else:
            status = NATIVE_WEB_SEARCH_STATUS_UNKNOWN
            reason = "compat_connection_unverified"
            source = "connection_inference"

        return {
            "provider": normalized_provider,
            "status": status,
            "reason": reason,
            "source": source,
            "official": official,
            "configured": configured,
            "supported": status == NATIVE_WEB_SEARCH_STATUS_SUPPORTED,
            "can_attempt": status
            in {
                NATIVE_WEB_SEARCH_STATUS_SUPPORTED,
                NATIVE_WEB_SEARCH_STATUS_UNKNOWN,
            },
            **({"connection_name": connection_name} if connection_name else {}),
        }

    if normalized_provider in {"google", "gemini"}:
        official = is_official_gemini_connection(url)
        if configured is True:
            status = NATIVE_WEB_SEARCH_STATUS_SUPPORTED
            reason = "connection_enabled"
            source = "connection_config"
        elif configured is False:
            status = NATIVE_WEB_SEARCH_STATUS_UNSUPPORTED
            reason = "connection_disabled"
            source = "connection_config"
        elif official:
            status = NATIVE_WEB_SEARCH_STATUS_SUPPORTED
            reason = "official_connection"
            source = "official_connection"
        else:
            status = NATIVE_WEB_SEARCH_STATUS_UNKNOWN
            reason = "compat_connection_unverified"
            source = "connection_inference"

        return {
            "provider": "gemini",
            "status": status,
            "reason": reason,
            "source": source,
            "official": official,
            "configured": configured,
            "supported": status == NATIVE_WEB_SEARCH_STATUS_SUPPORTED,
            "can_attempt": status
            in {
                NATIVE_WEB_SEARCH_STATUS_SUPPORTED,
                NATIVE_WEB_SEARCH_STATUS_UNKNOWN,
            },
            **({"connection_name": connection_name} if connection_name else {}),
        }

    return {
        "provider": normalized_provider or "unknown",
        "status": NATIVE_WEB_SEARCH_STATUS_UNSUPPORTED,
        "reason": "provider_not_supported",
        "source": "provider_capability",
        "official": False,
        "configured": configured,
        "supported": False,
        "can_attempt": False,
        **({"connection_name": connection_name} if connection_name else {}),
    }
